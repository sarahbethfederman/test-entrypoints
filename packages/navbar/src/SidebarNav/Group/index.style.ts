import styled from 'styled-components';
import { color } from '@lendi-ui/color';
import { ml, mt, mb } from '@lendi-ui/spacing';
import { Button } from '../Item/index.style';

export const Wrapper = styled.div``;

export interface ToggleProps {
  isExpanded: boolean;
}

export const MenuToggle = styled(Button)`
  display: flex;
  align-items: center;
  justify-content: space-between;
  svg {
    ${ml('xxs')} ${({ isExpanded }: ToggleProps) => isExpanded && `transform: rotate(-180deg);`};
  }
`;

export interface MenuProps {
  isExpanded: boolean;
}

// TODO: use height auto in transition - requires JS
// TODO: fix accessibility - e.g. menu children shouldn't tabbable when hidden
export const MenuContent = styled.div`
  max-height: 0;
  overflow: hidden;
  transition: max-height 0.25s;
  ${({ isExpanded }: MenuProps) => isExpanded && `max-height: 20em;`} > :first-child {
    ${mt('xxxs')};
  }

  > :last-child {
    ${mb('xs')};
  }

  border-bottom: 1px solid ${color('shade.100')};
`;
