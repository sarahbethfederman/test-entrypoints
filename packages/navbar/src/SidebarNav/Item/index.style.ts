import styled, { css } from 'styled-components';
import { px, py, mr } from '@lendi-ui/spacing';
import { body } from '@lendi-ui/typography';
import { bg, color, fg } from '@lendi-ui/color';

const buttonReset = css`
  ${py('nil')};
  border-top: none;
  border-right: none;
  border-left: none;
  background-color: transparent;
  width: 100%;
  outline: none;
`;

export interface StyleProps {
  depth?: number;
}

export const style = css`
  ${({ depth = 0 }: StyleProps) => {
    if (depth === 0) {
      return css`
        height: 3rem;
        ${body({ size: 'md', color: 'primary.500' })};
      `;
    } else {
      return css`
        height: 2.5rem;
        ${body({ size: 'sm', color: 'shade.700' })};
      `;
    }
  }} ${px('md')}
  font-weight: 600;
  cursor: pointer;
  display: flex;
  align-items: center;
  border-bottom: 1px solid ${color('shade.100')};

  :hover {
    ${fg('shade.0')} ${bg('primary.500')};
  }

  :active {
    ${fg('shade.0')} ${bg('primary.600')};
  }

  :last-of-type {
    border-bottom: none;
  }
`;

export const Link = styled.a`
  ${style};
  text-decoration: none;
`;

export const Button = styled.button`
  ${style} ${buttonReset};
`;

export const Icon = styled.div`
  ${mr('sm')};
`;
