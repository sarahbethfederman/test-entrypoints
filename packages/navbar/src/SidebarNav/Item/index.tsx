import * as React from 'react';
import { Link, Button, Icon, ListItem } from './index.style';

export interface ItemProps {
  href?: string;
  onClick?: () => void;
  depth?: number;
  icon?: React.ReactNode;
  children?: React.ReactNode;
  labelText?: string;
}

export const Item = (props: ItemProps) => {
  const { href, depth = 0, icon, children, onClick, labelText } = props;
  if (href) {
    return (
      <ListItem role="none">
        <Link depth={depth} href={href} role={labelText ? labelText : 'menu item'}>
          {icon && <Icon>{icon}</Icon>}
          {children}
        </Link>
      </ListItem>
    );
  } else {
    return (
      <ListItem role="none">
        <Button depth={depth} onClick={onClick} aria-label={labelText}>
          {icon && <Icon>{icon}</Icon>}
          {children}
        </Button>
      </ListItem>
    );
  }
};
