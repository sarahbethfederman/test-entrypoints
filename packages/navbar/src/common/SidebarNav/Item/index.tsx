import * as React from 'react';
import { Link, Button, Icon, ListItem } from './index.style';
import { AnalyticsContext } from '@lendi-ui/utils';
import { WindowPosition } from '@lendi/lendi-analytics-package';

export interface ItemProps {
  href?: string;
  onClick?: () => void;
  depth?: number;
  icon?: React.ReactNode;
  children?: React.ReactNode;
  labelText?: string;
}

export class Item extends React.Component<ItemProps, {}> {
  static contextType: any = AnalyticsContext;

  render() {
    const { href, depth = 0, icon, children, onClick = () => {}, labelText } = this.props;
    if (href) {
      return (
        <ListItem role="none">
          <Link
            depth={depth}
            href={href}
            role={labelText ? labelText : 'menu item'}
            onClick={() => {
              this.context.analyticsForNavigation(labelText ? labelText : 'menu item', WindowPosition.navigation_left);
            }}
          >
            {icon && <Icon>{icon}</Icon>}
            {children}
          </Link>
        </ListItem>
      );
    } else {
      return (
        <ListItem role="none">
          <Button
            depth={depth}
            onClick={() => {
              this.context.analyticsForNavigation(labelText!, WindowPosition.navigation_left);
              onClick();
            }}
            aria-label={labelText}
          >
            {icon && <Icon>{icon}</Icon>}
            {children}
          </Button>
        </ListItem>
      );
    }
  }
}
