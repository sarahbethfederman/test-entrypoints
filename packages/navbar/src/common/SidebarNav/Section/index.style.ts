import styled from 'styled-components';
import { bg } from '@lendi-ui/color';
import { px, py } from '@lendi-ui/spacing';
import { body } from '@lendi-ui/typography';

export const Title = styled.span`
  ${px('md')}
  ${py('xxs')}
  ${body({ size: 'xs' })}
  ${bg('shade.25')}
  display: block;
  font-weight: 600;
  text-transform: uppercase;
`;

export const ListWrapper = styled.ul`
  width: 100%;
  padding: 0;
  margin: 0;
`;
