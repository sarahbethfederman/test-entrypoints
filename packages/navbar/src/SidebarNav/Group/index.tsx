import * as React from 'react';
import { ChevronDown } from '@lendi-ui/icon';
import { ItemProps } from '../Item';
import { Wrapper, MenuToggle, MenuContent } from './index.style';

export interface GroupProps {
  title: React.ReactNode;
  children?: React.ReactElement<ItemProps> | React.ReactElement<ItemProps>[];
}

export interface GroupState {
  isExpanded: boolean;
}

export class Group extends React.Component<GroupProps, GroupState> {
  state: GroupState = {
    isExpanded: false,
  };

  handleClick = () => {
    this.setState((state) => ({
      isExpanded: !state.isExpanded,
    }));
  };

  render() {
    const { title, children } = this.props;
    const { isExpanded } = this.state;
    return (
      <Wrapper>
        <MenuToggle isExpanded={isExpanded} onClick={this.handleClick}>
          <span>{title}</span>
          <ChevronDown color="primary.500" />
        </MenuToggle>
        <MenuContent isExpanded={isExpanded}>
          {React.Children.map(children, (child) => {
            if (!React.isValidElement<ItemProps>(child)) {
              return child;
            }
            return React.cloneElement<ItemProps>(child, { depth: 1 });
          })}
        </MenuContent>
      </Wrapper>
    );
  }
}
