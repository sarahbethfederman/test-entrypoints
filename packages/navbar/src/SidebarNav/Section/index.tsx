import * as React from 'react';
import * as Fragment from 'react-dot-fragment';
import { ItemProps } from '../Item';
import { GroupProps } from '../Group';
import { Title, ListWrapper } from './index.style';

// Restrict to Item/Group elements and allow conditional components
type SectionChild =
  | React.ReactElement<ItemProps | GroupProps>
  | React.ReactText
  | React.ReactNode
  | boolean
  | null
  | undefined;

export interface SectionProps {
  title?: React.ReactNode;
  children?: SectionChild | SectionChild[];
}

export const Section = (props: SectionProps) => {
  const { title, children } = props;
  return (
    <Fragment>
      {title && <Title>{title}</Title>}
      <ListWrapper aria-role="menu">{children}</ListWrapper>
    </Fragment>
  );
};
