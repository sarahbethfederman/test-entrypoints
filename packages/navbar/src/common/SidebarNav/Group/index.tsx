import * as React from 'react';
import * as Fragment from 'react-dot-fragment';
import { ExpandMore } from '@lendi-ui/icon';
import { ItemProps } from '../Item';
import { MenuToggle, MenuContent, ListWrapper, ListItem } from './index.style';

export interface GroupProps {
  title: React.ReactNode;
  children?: React.ReactElement<ItemProps> | React.ReactElement<ItemProps>[];
}

export interface GroupState {
  isExpanded: boolean;
}

export class Group extends React.Component<GroupProps, GroupState> {
  state: GroupState = {
    isExpanded: false,
  };

  handleClick = () => {
    this.setState((state) => ({
      isExpanded: !state.isExpanded,
    }));
  };

  render() {
    const { title, children } = this.props;
    const { isExpanded } = this.state;
    return (
      <Fragment>
        <ListItem role="none">
          <MenuToggle
            isExpanded={isExpanded}
            onClick={this.handleClick}
            role="expand submenu"
            aria-expanded={isExpanded}
          >
            <span>{title}</span>
            <ExpandMore color="primary.500" />
          </MenuToggle>
          <MenuContent isExpanded={isExpanded}>
            <ListWrapper role="menu">
              {React.Children.map(children, (child) => {
                if (!React.isValidElement<ItemProps>(child)) {
                  return child;
                }
                return React.cloneElement<ItemProps>(child, { depth: 1 });
              })}
            </ListWrapper>
          </MenuContent>
        </ListItem>
      </Fragment>
    );
  }
}
