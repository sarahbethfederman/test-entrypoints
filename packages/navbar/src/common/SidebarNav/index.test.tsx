import * as React from 'react';
import { shallow } from 'enzyme';
import { Group } from './Group/index';
import { Item } from './Item/index';
import { Section } from './Section/index';
import { SidebarNav } from './index';
import { Wrapper } from './index.style';

describe('Group', () => {
  it('should mount', () => {
    const nav = shallow(
      <SidebarNav labelText="Primary navigation">
        <Section>
          <Group title="A">
            <Item>A.1</Item>
          </Group>
          <Item>B</Item>
          <Item>C</Item>
        </Section>
      </SidebarNav>
    );
    expect(nav).toMatchSnapshot();
    expect(nav.find(Wrapper)).toHaveLength(1);
  });

  it('should specify role when label is defined', () => {
    const nav = shallow(
      <SidebarNav labelText="Primary navigation">
        <Section>
          <Group title="A">
            <Item>A.1</Item>
          </Group>
          <Item>B</Item>
          <Item>C</Item>
        </Section>
      </SidebarNav>
    );
    expect(nav.find('[role]').exists()).toEqual(true);
  });
});
