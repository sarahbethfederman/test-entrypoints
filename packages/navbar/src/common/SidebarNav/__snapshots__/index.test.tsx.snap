// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Group should mount 1`] = `
<styled.nav
  role="Primary navigation"
>
  <Component>
    <Group
      title="A"
    >
      <Item>
        A.1
      </Item>
    </Group>
    <Item>
      B
    </Item>
    <Item>
      C
    </Item>
  </Component>
</styled.nav>
`;
